client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfillaParameters.kt

```diff
@@ -0,0 +1,105 @@
+package app.cash.backfila.client.misk.internal
+
+import app.cash.backfila.client.misk.Backfill
+import app.cash.backfila.client.misk.BackfillConfig
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeRequest
+import app.cash.backfila.protos.clientservice.PrepareBackfillRequest
+import app.cash.backfila.protos.clientservice.RunBatchRequest
+import app.cash.backfila.protos.service.Parameter
+import com.google.inject.TypeLiteral
+import com.squareup.moshi.Types
+import java.lang.reflect.ParameterizedType
+import kotlin.reflect.KClass
+import kotlin.reflect.KParameter
+import kotlin.reflect.full.memberProperties
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.jvm.jvmErasure
+import misk.exceptions.BadRequestException
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+
+internal inline fun <reified T : Any> getBackfilaParametersForBackfill(backfillClass: KClass<Backfill<*, *, T>>): List<Parameter> {
+  val parametersClass = findParametersClass(backfillClass)
+
+  // Validate that we can handle the parameters if they are specified.
+  val handledTypes = listOf(String::class, Int::class)
+  for (parameter in parametersClass.primaryConstructor!!.parameters) {
+    check(parameter.type.jvmErasure in handledTypes) {
+      "Parameter data class has member $parameter with unhandled type ${parameter.type.jvmErasure}"
+    }
+  }
+  return parametersClass.primaryConstructor?.parameters?.map {
+    Parameter.Builder().name(it.name).build()
+  }
+      ?: emptyList()
+}
+
+internal inline fun <reified T : Any> getBackfilaParameterMap(parameters: T): Map<String, ByteString> {
+  val parametersClass = T::class
+
+  val map = mutableMapOf<String, ByteString>()
+
+  for (property in parametersClass.memberProperties) {
+    map[property.name] = (property.getter.call(parameters) as String).encodeUtf8()
+  }
+  return map
+}
+
+internal inline fun <T : Any> findParametersClass(backfillClass: KClass<out Backfill<*, *, T>>): KClass<T> {
+  // Like MyBackfill.
+  val thisType = TypeLiteral.get(backfillClass.java)
+
+  // Like Backfill<E, Id<E>, MyDataClass>.
+  val supertype = thisType.getSupertype(Backfill::class.java).type as ParameterizedType
+
+  // Like MyDataClass
+  return (Types.getRawType(supertype.actualTypeArguments[2]) as Class<T>).kotlin
+}
+
+internal class BackfilaParametersOperator<T : Any>(
+  backfillClass: KClass<out Backfill<*, *, T>>
+) {
+  val parametersClass: KClass<T> = findParametersClass(backfillClass)
+
+  /** Constructor parameters to create a new T. */
+  val constructorParameters: List<KParameter>
+
+  init {
+    constructorParameters = parametersClass.primaryConstructor!!.parameters
+    // TODO check parameters
+  }
+
+  fun getBackfillConfig(request: PrepareBackfillRequest) = constructBackfillConfig(
```

Comment:
I'd keep this class only concerned with parameters

Reply:
Yeah good point. Removed